proposals/flexible-vectors/README.md

```diff
@@ -1,7 +1,1184 @@
+
+## Immediate LaneIdx operands
+
+As the vector length is not known at compile time, it does not make much sense to specify lane indices as immediates.
+
+Therefore, the type `i32` will be used to specify a lane index.
+The range of a lane index is no further constrained, and is intepreted modulo the actual vector length.
+
+If we impose an upper bound on the SIMD width, we can further constraint the range of lane indices.
+For instance, a maximum width of 2048 would allow to store a lane index in a 8-bit integer.
+Similarily, a maximum width of 524288 would allow to store a lane index in a 16-bit integer.
+The would not change the instruction encoding as they all take runtime lane indices which are always `i32`.
+
+## Special Operations
+
+### Vector length
+
+Returns the number of elements.
+Returned value will always be the same during the whole execution of the program.
+
+- `vec.v8.length -> i32`
+- `vec.v16.length -> i32`
+- `vec.v32.length -> i32`
+- `vec.v64.length -> i32`
+- `vec.v128.length -> i32`
+
+### Mask count
+
+Returns the number of active lanes of a `mask`
+
+- `vec.m8.count(m: vec.m8) -> i32`
+- `vec.m16.count(m: vec.m16) -> i32`
+- `vec.m32.count(m: vec.m32) -> i32`
+- `vec.m64.count(m: vec.m64) -> i32`
+- `vec.m128.count(m: vec.m128) -> i32`
+
+```python
+def mask.S.count(m):
+    result = 0
+    for i in range(mask.S.length):
+      if m[i]:
+          result += 1
+    return result
+```
+
+### Mask all active
+
+Returns a mask where all lanes are active
+
+- `vec.m8.all -> vec.m8`
+- `vec.m16.all -> vec.m16`
+- `vec.m32.all -> vec.m32`
+- `vec.m64.all -> vec.m64`
+- `vec.m128.all -> vec.m128`
+
+```python
+def mask.S.all():
+    result = mask.S.New()
+    for i in range(mask.S.length):
+        result[i] = 1
+    return result
+```
+
+### Mask None active
+
+Returns a mask where no lane is active
+
+- `vec.m8.none -> vec.m8`
+- `vec.m16.none -> vec.m16`
+- `vec.m32.none -> vec.m32`
+- `vec.m64.none -> vec.m64`
+- `vec.m128.none -> vec.m128`
+
+```python
+def mask.S.none():
+    result = mask.S.New()
+    for i in range(mask.S.length):
+        result[i] = 0
+    return result
+```
+
+### Mask index CMP
+
+Returns a mask whose active lanes satisfy `(x + laneIdx) CMP n`
+CMP one of the following: `eq`, `ne`, `lt`, `le`, `gt`, `ge`
+
+- `vec.m8.index_CMP(x: i32, n: i32) -> vec.m8`
+- `vec.m16.index_CMP(x: i32, n: i32) -> vec.m16`
+- `vec.m32.index_CMP(x: i32, n: i32) -> vec.m32`
+- `vec.m64.index_CMP(x: i32, n: i32) -> vec.m64`
+- `vec.m128.index_CMP(x: i32, n: i32) -> vec.m128`
+
+```python
+def vec.S.index_CMP(x, n):
+    result = vec.S.New()
+    for i in range(vec.S.length):
+        if (x + i) CMP n:
+            result[i] = 1
+        else:
+            result[i] = 0
+    return result
+```
+
+### Mask index first
+
+Returns the index of the first active lane.
+If there is no active lane, the index of the last lane is returned.
```

Comment:
Shouldn't the number of lanes (i.e. the index of the last lane + 1) be returned instead? In this way we can check if there is an active lane.

Reply:
I changed it `last lane + 1`, ie: vec length